### ISSUE TITLE
Evaluation results inconsistency outputted by train.py(set test=true) and eval.py on the best ckpt.

### ISSUE THREAD

**Issue by zcx-language** [ashleve/lightning-hydra-template#527]
![image](https://user-images.githubusercontent.com/38426607/218422816-83cfba48-b3e1-42a0-9421-2d8fff561ab9.png)

The image shows the code snippet in `train.py`, while I use a pre-trained ckpt to evaluate, and the model, datamodule, trainer are re-instantiated.

When running `train.py`, if I set `train=false`  in `train.yaml` (i.e. not execute the code `trainer.fit(...)` in the red rectangle), it outputs the same evaluation results every time. However,  if I set `train=true`, it outputs different results every time. I want to kown if `trainer.fit(...)` changed some global things that make the random state change?? Thanks.

Note that the test_step() uses a randomized operation, but I have set a random seed in the `train.yaml`.

**Comment by ashleve** [comment]
>  I want to kown if trainer.fit(...) changed some global things that make the random state change?

I don't think that's the case. Perhaps the way you load your model from ckpt is random for some reason?

I struggle to replicate you results. I did the following:
1. Downloaded template `main` branch
1. Set `seed=123` in `train.yaml`
2. Changed the code to:
```python
    if cfg.get("train"):
        log.info("Starting training!")
        trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.get("ckpt_path"))

    train_metrics = trainer.callback_metrics

    if cfg.get("test"):

        model: LightningModule = hydra.utils.instantiate(cfg.model)
        datamodule: LightningDataModule = hydra.utils.instantiate(cfg.data)
        trainer: Trainer = hydra.utils.instantiate(cfg.trainer)

        trainer.test(model=model, datamodule=datamodule, ckpt_path="last.ckpt") # the ckpt I trained before
```

And the following test result table is the same on every run:
<img width="344" alt="image" src="https://user-images.githubusercontent.com/43458894/218464793-bec56ebd-b3d0-4f7d-983e-207b0177c3d7.png">

**Comment by zcx-language** [comment]
Thanks, I tried the original mnist example and its outputs are normal. 
> Perhaps the way you load your model from ckpt is random for some reason?

In my project, the ‘pretrained.ckpt’ is an absolute path, so I think the loaded ckpt should be the same every time. The difference only lay on the train=true or false. 

![image](https://user-images.githubusercontent.com/38426607/218486370-8fcd6fb2-b097-40df-b18b-be0213e8c488.png)

I change the original mnist example a little, and the same problem occurred. As the image shows above, by adding random noise to the inputs (though it makes no sense), the test results outputted by `train.py` is different from the results outputted by `eval.py` with the same best ckpt.  Even setting the same seed in `train.yaml` and `eval.yaml`, they also output different results.

When I set  `trainer.max_epochs=3`  and `seed=42` in both `train.yaml` and `eval.yaml`, and run on a gpu device, the `train.py` outputs:
![image](https://user-images.githubusercontent.com/38426607/218492555-531af315-d7be-44e0-9c2d-cd43483bbc1f.png)
while the eval.py outputs:
![image](https://user-images.githubusercontent.com/38426607/218493692-0be09a9c-f7de-4617-8b3e-e569ee9c226f.png)

**Comment by ashleve** [comment]
So if I understand correctly, you're adding random noise at inference? This sounds like the problem with seeding the noise generation. I would check if noise is the same on every run (I don't know how you generate it but there's probably some source of randomness that `pl.seed_everything()` doesn't seed)

**Comment by zcx-language** [comment]
> This sounds like the problem with seeding the noise generation. I would check if noise is the same on every run

Maybe you are right, but I think it is more complicated. When I run eval.py with the same ckpt and seed multiple times, they output the same result every time (I.e. torch.randn is seeded). But the result is unequal to the test result outputted by train.py even it test on the same ckpt with same seed.

**Comment by zcx-language** [comment]
I think I have grasped the underlying cause.

Despite the random seed being set, the generation of random numbers in `train.py` and `eval.py` differ from each other. Random numbers are generated during the execution of trainer.fit() in `train.py`, which influences the subsequent generation of random numbers during `trainer.test()`.

As exemplified below:
![image](https://user-images.githubusercontent.com/38426607/218624491-ba01b58a-391b-4b2c-8d9b-1ffa50795f5a.png)
![image](https://user-images.githubusercontent.com/38426607/218624575-69d16062-6c8b-4c76-a24d-a66ed0a3c5ca.png)

**Comment by zcx-language** [comment]
Resetting the random seed before executing `trainer.test()` will ensure that the fit process does not impact the test process.
